internal/impl/kafka/enterprise/schema_registry_input.go

```diff
@@ -0,0 +1,265 @@
+		Beta().
+		Version("4.32.2").
+		Categories("Integration").
+		Summary(`Reads schemas from SchemaRegistry.`).
+		Description(`
+== Metadata
+
+This input adds the following metadata fields to each message:
+
+`+"```text"+`
+- schema_registry_subject
+- schema_registry_version
+`+"```"+`
+
+You can access these metadata fields using
+xref:configuration:interpolation.adoc#bloblang-queries[function interpolation].
+
+`).
+		Fields(
+			service.NewStringField(sriFieldURL).Description("The base URL of the schema registry service."),
+			service.NewBoolField(sriFieldIncludeDeleted).Description("Include deleted entities.").Default(false).Advanced(),
+			service.NewStringField(sriFieldSubjectFilter).Description("Include only subjects which match the regular expression filter. All subjects are selected when not set.").Default("").Advanced(),
+			service.NewTLSToggledField(sriFieldTLS),
+			service.NewAutoRetryNacksToggleField(),
+		)
+}
+
+func init() {
+	err := service.RegisterInput("schema_registry", inputSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.Input, error) {
+			i, err := inputFromParsed(conf, mgr.Logger())
+			if err != nil {
+				return nil, err
+			}
+			return service.AutoRetryNacksToggled(conf, i)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+type input struct {
+	url           *url.URL
+	subjectFilter *regexp.Regexp
+
+	client    http.Client
+	connected atomic.Bool
+	subjects  []string
+	subject   string
+	versions  []int
+	log       *service.Logger
+}
+
+func inputFromParsed(pConf *service.ParsedConfig, log *service.Logger) (i *input, err error) {
+	i = &input{
+		log: log,
+	}
+
+	var u string
+	if u, err = pConf.FieldString(sriFieldURL); err != nil {
+		return
+	}
+	if i.url, err = url.Parse(u); err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %s", err)
+	}
+
+	var includeDeleted bool
+	if includeDeleted, err = pConf.FieldBool(sriFieldIncludeDeleted); err != nil {
+		return
+	}
+	if includeDeleted {
+		q := i.url.Query()
+		q.Add("deleted", "true")
+		i.url.RawQuery = q.Encode()
+	}
+
+	var filter string
+	if filter, err = pConf.FieldString(sriFieldSubjectFilter); err != nil {
+		return
+	}
+	if i.subjectFilter, err = regexp.Compile(filter); err != nil {
+		return nil, fmt.Errorf("failed to compile subject filter %q: %s", filter, err)
+	}
+
+	var tlsConf *tls.Config
+	var tlsEnabled bool
+	if tlsConf, tlsEnabled, err = pConf.FieldTLSToggled(sriFieldTLS); err != nil {
+		return
+	}
+
+	i.client = http.Client{}
+	if tlsEnabled && tlsConf != nil {
+		if c, ok := http.DefaultTransport.(*http.Transport); ok {
+			cloned := c.Clone()
+			cloned.TLSClientConfig = tlsConf
+			i.client.Transport = cloned
+		} else {
+			i.client.Transport = &http.Transport{
+				TLSClientConfig: tlsConf,
+			}
+		}
+	}
+
+	return
+}
+
+//------------------------------------------------------------------------------
+
+func (i *input) Connect(ctx context.Context) error {
+	if i.connected.Load() {
```

Comment:
Shouldn't this be a lock otherwise we'll get races writing to `i.subjects`?

Reply:
I just noticed there's a new `atomic.Bool` thing since Go 1.19 and thought it would be fun to play with it, but you're right and I should've put more thought into this. Oh well... Locks it is :) Thank you for catching this! ❤️